### PR TITLE
[modular] wirebrush

### PR DESCRIPTION
**Pull request by jjpark-kb**
## About The Pull Request

adds a wirebrush (and an advanced version that is faster) that "removes" (read: changeturf) rusted walls and plating.

## Why It's Good For The Game

janitors being worth more, cleaning a station makes it look nicer = less chaos.

## Changelog
:cl:
add: added two wirebrushes (basic and advanced)
soundadd: added a wirebrush sound
imageadd: added two wirebrush icons (basic and advanced)
/:cl:

### CODE REVIEW

File: modular_skyrat/modules/wirebrush_rust/code/game/objects/items/wirebrush.dm [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	icon = 'modular_skyrat/modules/wirebrush_rust/icons/obj/wirebrush.dmi'
+	icon_state = "wirebrush"
+	///variable that determines how long it takes to "clean" (read: replace(?)) rusted walls.
+	var/clean_speed = 30 //3 seconds
+
+/obj/item/wirebrush/afterattack(atom/target, mob/user, proximity_flag, click_parameters)
+	. = ..()
+	if(istype(target, /turf/open/floor/plating/rust))
+		if(!do_after(user, clean_speed, target = target))
+			to_chat(user, span_warning("You stop cleaning, leaving the wall still rusty!"))
```
Comment (tf-4) [inline modular_skyrat/modules/wirebrush_rust/code/game/objects/items/wirebrush.dm]:
```suggestion
			to_chat(user, span_warning("You stop cleaning, leaving the plating still rusty!"))
```

File: modular_skyrat/modules/wirebrush_rust/code/modules/food_and_drinks/kitchen_machinery/microwave.dm [hunk 2]
```diff
@@ -0,0 +1,99 @@
+/obj/machinery/microwave/attackby(obj/item/O, mob/living/user, params)
```
Comment (tf-4) [inline modular_skyrat/modules/wirebrush_rust/code/modules/food_and_drinks/kitchen_machinery/microwave.dm:1]: Use descriptive variable names

File: modular_skyrat/modules/wirebrush_rust/code/game/objects/items/wirebrush.dm [hunk 3]
```diff
@@ -0,0 +1,45 @@
+/obj/item/wirebrush
+	name = "wirebrush"
+	desc = "A tool that is used to scrub the rust thoroughly off walls. Not for hair or fur!"
+	icon = 'modular_skyrat/modules/wirebrush_rust/icons/obj/wirebrush.dmi'
+	icon_state = "wirebrush"
+	///variable that determines how long it takes to "clean" (read: replace(?)) rusted walls.
+	var/clean_speed = 30 //3 seconds
+
+/obj/item/wirebrush/afterattack(atom/target, mob/user, proximity_flag, click_parameters)
+	. = ..()
```
Comment (ghost) [inline modular_skyrat/modules/wirebrush_rust/code/game/objects/items/wirebrush.dm:20]: optionally, you could make this work with the cleaning skill similar to other janitor tools! optional tho
```suggestion
	. = ..()
	if(user.mind)
		cleanspeed = cleanspeed * min(user.mind.get_skill_modifier(/datum/skill/cleaning, SKILL_SPEED_MODIFIER)+0.1,1)
```

File: code/modules/research/techweb/all_nodes.dm [hunk 4]
```diff
@@ -1220,6 +1221,10 @@
 		"light_replacer",
 		"paint_remover",
 		"spraybottle",
```
Comment (ghost) [inline code/modules/research/techweb/all_nodes.dm:1223]: extra space reeee
```suggestion
		"spraybottle",
		//SKYRAT EDIT START - RESEARCH DESIGNS
```

File: modular_skyrat/modules/wirebrush_rust/code/game/objects/items/wirebrush.dm (hunk 1 above)
Comment (ghost) [reply modular_skyrat/modules/wirebrush_rust/code/game/objects/items/wirebrush.dm]: alternatively, you could create the rustedTurf var before the do_after, and then just use [rustedTurf]

File: modular_skyrat/modules/wirebrush_rust/code/modules/food_and_drinks/kitchen_machinery/microwave.dm [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+			return TRUE
+		if(!user.transferItemToLoc(O, src))
+			to_chat(user, span_warning("\The [O] is stuck to your hand!"))
+			return FALSE
+
+		ingredients += O
+		user.visible_message(span_notice("[user] adds \a [O] to \the [src]."), span_notice("You add [O] to \the [src]."))
+		return
+
+	..()
```
Comment (ghost) [inline modular_skyrat/modules/wirebrush_rust/code/modules/food_and_drinks/kitchen_machinery/microwave.dm:99]: I dont think this entire proc needs re-written.? could you not use this to add the wirebrush function to the proc? or did you try that and it was fucky
```suggestion
/obj/machinery/microwave/attackby(obj/item/O, mob/living/user, params)
	. = ..()
	if(istype(O, /obj/item/wirebrush))
		var/obj/item/wirebrush/usedObj = O
		user.visible_message(span_notice("[user] starts to clean \the [src]."), span_notice("You start to clean \the [src]..."))
		if(do_after(user, usedObj.clean_speed, target = src))
			user.visible_message(span_notice("[user] cleans \the [src]."), span_notice("You clean \the [src]."))
			dirty = 0
			update_appearance()
		return TRUE
```

Comment (ghost) [reply modular_skyrat/modules/wirebrush_rust/code/modules/food_and_drinks/kitchen_machinery/microwave.dm:99]: you would also then be able to put this into the wirebrush's item file if you wanted (if this worked)

Comment (jjpark-kb) [reply modular_skyrat/modules/wirebrush_rust/code/modules/food_and_drinks/kitchen_machinery/microwave.dm:99]: modularization was the intent.

File: modular_skyrat/modules/wirebrush_rust/code/modules/food_and_drinks/kitchen_machinery/microwave.dm (hunk 2 above)
Comment (jjpark-kb) [reply modular_skyrat/modules/wirebrush_rust/code/modules/food_and_drinks/kitchen_machinery/microwave.dm:1]: this is just modularization, a copy paste of the core code.

File: modular_skyrat/modules/wirebrush_rust/code/game/objects/items/wirebrush.dm (hunk 1 above)
Comment (jjpark-kb) [reply modular_skyrat/modules/wirebrush_rust/code/game/objects/items/wirebrush.dm]: solved this issue @tf-4

File: code/modules/research/techweb/all_nodes.dm (hunk 4 above)
Comment (jjpark-kb) [reply code/modules/research/techweb/all_nodes.dm:1223]: the space is how its done in many other instances within this file.

File: modular_skyrat/modules/wirebrush_rust/code/modules/food_and_drinks/kitchen_machinery/microwave.dm (hunk 5 above)
Comment (ghost) [reply modular_skyrat/modules/wirebrush_rust/code/modules/food_and_drinks/kitchen_machinery/microwave.dm:99]: that'd still be modular tho 🤔 it wouldnt overwrite anything, just append the new code to the end of the old stuff

**Comment by SpaceVampire** [comment]
Massive stealth nerf to rust-path Heretics.

File: modular_skyrat/modules/wirebrush_rust/code/game/objects/items/wirebrush.dm [hunk 6]
```diff
@@ -0,0 +1,39 @@
+/obj/item/wirebrush
+	name = "wirebrush"
+	desc = "A tool that is used to scrub the rust thoroughly off walls. Not for hair or fur!"
+	icon = 'modular_skyrat/modules/wirebrush_rust/icons/obj/wirebrush.dmi'
+	icon_state = "wirebrush"
+	///variable that determines how long it takes to "clean" (read: replace(?)) rusted walls.
+	var/clean_speed = 30 //3 seconds
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/wirebrush_rust/code/game/objects/items/wirebrush.dm:7]:
```suggestion
	var/clean_speed = 3 SECONDS
```

**Comment by ghost** [comment]
so, while i was looking for something totally unrelated, I found a proc that might come in handy for this actually


```
/turf/proc/ScrapeAway(amount=1, flags)
	if(!amount)
		return
	if(length(baseturfs))
		var/list/new_baseturfs = baseturfs.Copy()
		var/turf_type = new_baseturfs[max(1, new_baseturfs.len - amount + 1)]
		while(ispath(turf_type, /turf/baseturf_skipover))
			amount++
			if(amount > new_baseturfs.len)
				CRASH("The bottommost baseturf of a turf is a skipover [src]([type])")
			turf_type = new_baseturfs[max(1, new_baseturfs.len - amount + 1)]
		new_baseturfs.len -= min(amount, new_baseturfs.len - 1) // No removing the very bottom
		if(new_baseturfs.len == 1)
			new_baseturfs = new_baseturfs[1]
		return ChangeTurf(turf_type, new_baseturfs, flags)

	if(baseturfs == type)
		return src

	return ChangeTurf(baseturfs, baseturfs, flags) // The bottom baseturf will never go away
```

**Comment by jjpark-kb** [comment]
> so, while i was looking for something totally unrelated, I found a proc that might come in handy for this actually
> 
> ```
> /turf/proc/ScrapeAway(amount=1, flags)
> 	if(!amount)
> 		return
> 	if(length(baseturfs))
> 		var/list/new_baseturfs = baseturfs.Copy()
> 		var/turf_type = new_baseturfs[max(1, new_baseturfs.len - amount + 1)]
> 		while(ispath(turf_type, /turf/baseturf_skipover))
> 			amount++
> 			if(amount > new_baseturfs.len)
> 				CRASH("The bottommost baseturf of a turf is a skipover [src]([type])")
> 			turf_type = new_baseturfs[max(1, new_baseturfs.len - amount + 1)]
> 		new_baseturfs.len -= min(amount, new_baseturfs.len - 1) // No removing the very bottom
> 		if(new_baseturfs.len == 1)
> 			new_baseturfs = new_baseturfs[1]
> 		return ChangeTurf(turf_type, new_baseturfs, flags)
> 
> 	if(baseturfs == type)
> 		return src
> 
> 	return ChangeTurf(baseturfs, baseturfs, flags) // The bottom baseturf will never go away
> ```

scrapeaway is for removing the turf. from wall to plating, as it says the baseturf.

**Comment by EvilCrimsonWitch** [comment]
I still don't think these work, unfortunately. Tried using them today on rusted walls and it kept telling me I could only use them on rusted walls and floors.

**Pull request merged**